micro_sam/evaluation/inference.py

```diff
@@ -392,118 +389,127 @@ def _save_segmentation(masks, prediction_path):
-            }]
-
-            sampled_binary_y = torch.stack([
-                torch.stack([_gt == idx for idx in sampled[batch_start:batch_stop]])[:, None]
-                for _gt, sampled in zip(gt, sampled_ids)
-            ]).to(torch.float32)
-
-            batched_outputs = model(
-                this_batched_inputs,
-                multimask_output=multimasking if iteration == 0 else False,
-                image_embeddings=image_embeddings
-            )
-
-            masks, logits_masks = [], []
-            for m in batched_outputs:
-                mask, l_mask = [], []
-                for _m, _l, _iou in zip(m["masks"], m["low_res_masks"], m["iou_predictions"]):
-                    best_iou_idx = torch.argmax(_iou)
-                    mask.append(torch.sigmoid(_m[best_iou_idx][None]))
-                    l_mask.append(_l[best_iou_idx][None])
-                mask, l_mask = torch.stack(mask), torch.stack(l_mask)
-                masks.append(mask)
-                logits_masks.append(l_mask)
+        batched_outputs = batched_inference(
+            predictor, image, batch_size,
+            boxes=boxes, points=points, point_labels=point_labels,
+            multimasking=multimasking, embedding_path=embedding_path,
+            return_instance_segmentation=False
+        )
+        masks = extract_instances_from_batched_outputs(batched_outputs)
 
-            masks, logits_masks = torch.stack(masks), torch.stack(logits_masks)
-            masks = (masks > 0.5).to(torch.float32)
-            final_masks.append(masks)
+        next_coords, next_labels, _, _ = prompt_generator(sampled_binary_gt, masks)
+        next_coords, next_labels = next_coords.detach().cpu().numpy(), next_labels.detach().cpu().numpy()
 
-            for _pred, _gt, _inp, logits in zip(masks, sampled_binary_y, this_batched_inputs, logits_masks):
-                next_coords, next_labels, _, _ = prompt_generator(_gt, _pred)
-                updated_point_coords = torch.cat([_inp["point_coords"], next_coords], dim=1) \
-                    if "point_coords" in _inp.keys() else next_coords
-                updated_point_labels = torch.cat([_inp["point_labels"], next_labels], dim=1) \
-                    if "point_labels" in _inp.keys() else next_labels
+        if points is not None:
+            points = np.concatenate([points, next_coords], axis=1)
+        else:
+            points = next_coords
 
-                _inp["point_coords"] = updated_point_coords
-                _inp["point_labels"] = updated_point_labels
-                _inp["mask_inputs"] = logits
+        if point_labels is not None:
+            point_labels = np.concatenate([point_labels, next_labels], axis=1)
+        else:
+            point_labels = next_labels
 
-        final_masks = torch.cat(final_masks, dim=1)
-        _save_segmentation(final_masks, prediction_paths[iteration])
+        _save_segmentation(masks, prediction_paths[iteration])
```

Comment:
Set `multimasking = False` at the end of the loop. Otherwise it will stay true if it was set to true in the beginning. But that is wrong because we have more than one prompt after the first iteration.

Reply:
Okay this is crucial. Thanks for pointing it out.